persistent/picklecache.py

```diff
@@ -62,6 +65,87 @@ def locked(self, *args, **kwargs):
     return locked
 
 
+class _WeakValueDictionary(object):
+    # Maps from OID -> Persistent object, but
+    # only weakly references the Persistent object. This is similar
+    # to ``weakref.WeakValueDictionary``, but is customized depending on the
+    # platform. On PyPy, all objects can cheaply use a WeakRef, so that's
+    # what we actually use. On CPython, though, ``PersistentPy`` cannot be weakly
+    # referenced, so we rely on the fact that the ``id()`` of an object is its
+    # memory location, and we use ``ctypes`` to cast that integer back to
+    # the object.
+    #
+    # To remove stale addresses, we rely on the ``ffi.gc()`` object with the exact
+    # same lifetime as the ``PersistentPy`` object. It calls us, we get the ``id``
+    # back out of the CData, and clean up.
+    if PYPY: # pragma: no cover
+        def __init__(self):
+            self._data = WeakValueDictionary()
+
+        def _from_addr(self, addr):
+            return addr
+
+        def _save_addr(self, oid, obj):
+            return obj
+
+        cleanup_hook = None
+    else:
+        def __init__(self):
+            self._data = {}
+            self._addr_to_oid = {}
+
+        def _save_addr(self, oid, obj):
+            i = id(obj)
+            self._addr_to_oid[i] = oid
+            return i
+
+        try:
+            from ctypes import cast
+            from ctypes import py_object
+            cast = staticmethod(cast)
+        except ImportError: # pragma: no cover
+            def _from_addr(self, addr):
+                for i in gc.get_objects():
+                    if id(i) == addr:
+                        return i
```

Comment:
I don't like this.  id() values can be reused for unrelated objects in CPython.

Reply:
Very true! 

I don't expect this to be used on CPython. Even if it were, the ref counting semantics that are tied to `ffi.gc` should eliminate the race condition.

Reply:
That fallback loop was borrowed from `objgraph`, IIRC. Perhaps the state of the art for non-CPython implementations has moved on since I checked?

Reply:
objgraph is a debugging tool.  I allow hacks in debugging code that I wouldn't allow in production code ;)

Reply:
I think my primary reason for rejecting this is that AFAIU `_from_addr()` is supposed to return None if the object associated with the old id() value has been freed, but the get_objects() method cannot reliably implement that.